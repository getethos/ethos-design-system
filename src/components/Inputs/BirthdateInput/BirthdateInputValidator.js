import React from 'react'
import dayjs from '../../../helpers/getDayjs.js'

export const DATE_FORMATS = [
  'mm/dd/yyyy',
  'mm/yyyy',
  'mm/yy',
]

export const dateMaskByFormat = {
  'mm/dd/yyyy': [/\d/, /\d/, '/', /\d/, /\d/, '/', /\d/, /\d/, /\d/, /\d/],
  'mm/yyyy': [/\d/, /\d/, '/', /\d/, /\d/, /\d/, /\d/],
  'mm/yy': [/\d/, /\d/, '/', /\d/, /\d/],
}

export const dateRegexByFormat = {
  'mm/dd/yyyy': /\d\d\/\d\d\/\d\d\d\d/,
  'mm/yyyy': /\d\d\/\d\d\d\d/,
  'mm/yy': /\d\d\/\d\d/,
}

// first replace removes text-mask's underscores, second fixes like: '2//'
export const cleanse = (value) => value.replace(/[_]/g,'').replace('//', '/')

export const dateStringMatchesFormat = (cleansedDateString, dateFormat) => {
  const pattern = dateRegexByFormat[dateFormat];
  const matchesFormat = pattern.test(cleansedDateString);
  if (!matchesFormat) {
    return 'Please enter a valid date.'
  }
  return '';
}

export const getMaxDateValidator = (props) => {
  const {
    maxDate,
    customErrorMessage,
    dateFormat,
  } = props

  return (value) => {
    if (value == null || value === '') return ''
    const date = dayjs(maxDate);
    const dayjsFormat = dateFormat.toUpperCase();
    const errorMessage =
      customErrorMessage ||
      `Please enter a date before ${date.format(dayjsFormat)}`
    return dayjs(value).isBefore(date) ? '' : errorMessage
  }
}

export const getMinDateValidator = (props) => {
  const {
    minDate,
    customErrorMessage,
    dateFormat,
  } = props

  return (value) => {
    // Should pass if there is no value
    if (value == null || value === '') return ''
    const date = dayjs(minDate);
    const dayjsFormat = dateFormat.toUpperCase();
    const errorMessage =
      customErrorMessage ||
      `Please enter a date before ${date.format(dayjsFormat)}`
    return dayjs(value).isAfter(date) ? '' : errorMessage
  }
}

export const getMinMaxDateValidator = (props) => {
  const {
    minAge,
    maxAge,
    dateFormat,
    customErrorMessage,
  } = props;

  return (value) => {
    const dateRangeErrorMessage = customErrorMessage || `Sorry, you must be ${minAge}â€“${maxAge}.`
    const minBirthdate = dayjs()
      .subtract(maxAge + 1, 'years')
      .endOf('day')
      .toDate()

    const maxBirthdate = dayjs()
      .subtract(minAge, 'years')
      .endOf('day')
      .toDate()

    const maxError = getMaxDateValidator({
      maxDate: maxBirthdate,
      customErrorMessage: dateRangeErrorMessage,
      dateFormat,
    })(value);

    const minError = getMinDateValidator({
      minDate: minBirthdate,
      customErrorMessage: dateRangeErrorMessage,
      dateFormat,
    })(value);

    return minError.length || maxError.length ? dateRangeErrorMessage : '';
  }
}
