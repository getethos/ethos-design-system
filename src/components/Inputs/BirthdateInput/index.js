export { BirthdateInput } from './BirthdateInput.js'
export { default as BirthdateInputValidator } from './BirthdateInputValidator.js'
