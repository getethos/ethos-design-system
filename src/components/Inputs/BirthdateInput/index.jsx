export { BirthdateInput } from './BirthdateInput'
export { default as BirthdateInputValidator } from './BirthdateInputValidator'
