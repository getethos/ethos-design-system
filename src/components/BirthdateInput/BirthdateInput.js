import React from 'react'
import PropTypes from 'prop-types'
import { DateInput } from '../DateInput'
import * as Validators from '../../validators/BirthdateInputValidator'
const { DATE_FORMATS } = Validators

export const BirthdateInput = (props) => {
  const {
    optional,
    dateFormat,
    allCaps,
    capitalize,
    labelCopy,
    validator,
    formChangeHandler,
    initialValue,
    currentValue,
    currentError,
    formTouched,
    setFieldTouched,
    autoComplete,
    classOverrides,
    ...restProps
  } = props
  return (
    <DateInput
      optional={optional}
      dateFormat={dateFormat}
      allCaps={allCaps}
      capitalize={capitalize}
      labelCopy={labelCopy}
      validator={validator}
      formChangeHandler={formChangeHandler}
      initialValue={initialValue}
      currentValue={currentValue}
      currentError={currentError}
      formTouched={formTouched}
      setFieldTouched={setFieldTouched}
      autoComplete={autoComplete}
      classOverrides={classOverrides}
      {...restProps}
    />
  )
}

BirthdateInput.propTypes = {
  /** whether this field is optional or not */
  optional: PropTypes.bool,
  /** Date format -- one of: 'mm/dd/yyyy', 'mm/yyyy', 'mm/yy' */
  dateFormat: PropTypes.oneOf(DATE_FORMATS),
  /** Required data-tid used as a unique id for targeting test selectors */
  'data-tid': PropTypes.string,
  /** whether disabled or not */
  disabled: PropTypes.bool,
  /** whether to display label in all caps */
  allCaps: PropTypes.bool,
  /** text transform capitalize label */
  capitalize: PropTypes.bool,
  /** Name of the field */
  name: PropTypes.string.isRequired,
  /** label  */
  labelCopy: PropTypes.string,
  /** field validation callback */
  validator: PropTypes.func,
  /** Optionally sets a default value for the birthdate */
  initialValue: PropTypes.string,
  /** low level prop used only by the form engine */
  formTouched: PropTypes.bool,
  /** currentValue is a low level prop used only by the form engine */
  currentValue: PropTypes.string,
  /** setFieldTouched is a low level prop used only by the form engine */
  setFieldTouched: PropTypes.func,
  /** currentError is a low level prop used only by the form engine */
  currentError: PropTypes.string,
  /** formChangeHandler is a low level prop used only by the form engine */
  formChangeHandler: PropTypes.func,
  /** autoComplete value for auto fill */
  autoComplete: PropTypes.string,
  /** passed down through component to override Inputstyles */
  classOverrides: PropTypes.string,
}

export const BirthdateInputValidators = Validators
