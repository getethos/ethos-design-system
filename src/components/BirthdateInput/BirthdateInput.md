The BirthdateInput takes a validator, but will also internally validate that the
date string corresponds with `dateFormat` (once the field has been `touched` similar
to [redux-form](https://redux-form.com/8.2.2/docs/api/field.md/#2-a-stateless-function))

Important to note that `getMinBirthdate` should take the argument of `maximumEligibleAge` and
`getMaxBirthdate` takes the argument of `minimumEligibleAge`

```jsx
import { validateMinMaxDateFactory } from '../../validators/BirthdateInputValidator'
import dayjs from '../../helpers/getDayjs.js'

const minimumEligibleAge = 20
const maximumEligibleAge = 65

const getMinBirthdateLga = (maxAge) => {
  // Max age determines the earliest (minimum) allowable birthdate
  const minBirthdate = dayjs()
    .subtract(maxAge, 'years')
    .subtract(6, 'months')
    .add(1, 'days')
    .startOf('day')
    .toDate()
  return minBirthdate
}

const getMaxBirthdateLga = (minAge) => {
  // Min age determines the latest (maximum) allowable birthdate
  const maxBirthdate = dayjs()
    .subtract(minAge, 'years')
    .add(6, 'months')
    .subtract(1, 'day')
    .endOf('day')
    .toDate()
  return maxBirthdate
}

// formChangeHandler gets wired up automatically if using <Form /> component
const formChangeHandlerStub = () => {}

;<BirthdateInput
  name="birthDate"
  labelCopy="Birthdate"
  allCaps={true}
  data-tid="the-birthdate-input"
  formChangeHandler={formChangeHandlerStub}
  validator={(value) => {
    return validateMinMaxDateFactory({
      minAge: minimumEligibleAge,
      maxAge: maximumEligibleAge,
      minBirthdate: getMinBirthdateLga(maximumEligibleAge),
      maxBirthdate: getMaxBirthdateLga(minimumEligibleAge),
      dateFormat: 'mm/dd/yyyy',
    })(value)
  }}
/>
```

This one sets an `intialValue` which results in the field being considered as
already `touched`. This means you do not have to `blur` for field hint error
messages to appear.

```jsx
import { validateMinMaxDateFactory } from '../../validators/BirthdateInputValidator'
import dayjs from '../../helpers/getDayjs.js'

const minimumEligibleAge = 20
const maximumEligibleAge = 65

const getMinBirthdateLga = (maxAge) => {
  // Max age determines the earliest (minimum) allowable birthdate
  const minBirthdate = dayjs()
    .subtract(maxAge, 'years')
    .subtract(6, 'months')
    .add(1, 'days')
    .startOf('day')
    .toDate()
  return minBirthdate
}

const getMaxBirthdateLga = (minAge) => {
  // Min age determines the latest (maximum) allowable birthdate
  const maxBirthdate = dayjs()
    .subtract(minAge, 'years')
    .add(6, 'months')
    .subtract(1, 'day')
    .endOf('day')
    .toDate()
  return maxBirthdate
}

// formChangeHandler gets wired up automatically if using <Form /> component
const formChangeHandlerStub = () => {}

;<BirthdateInput
  initialValue="09/12/"
  name="birthDate"
  labelCopy="Birthdate"
  allCaps={true}
  data-tid="the-birthdate-input"
  formChangeHandler={formChangeHandlerStub}
  validator={(value) => {
    return validateMinMaxDateFactory({
      minAge: minimumEligibleAge,
      maxAge: maximumEligibleAge,
      minBirthdate: getMinBirthdateLga(maximumEligibleAge),
      maxBirthdate: getMaxBirthdateLga(minimumEligibleAge),
      dateFormat: 'mm/dd/yyyy',
    })(value)
  }}
/>
```

BirthdateInput used as a single field form that's prefilled

```jsx
import validateExists from '../../validators/validateExists'

import {
  TitleLarge,
  Form,
  Spacer,
  Button,
  InfoMessage,
} from '../index'

;<Form
  config={{
    formName: 'Prefilled w/Birthdate',
    formId: '1',
    fields: {
      birthdate: {
        component: (props, options) => {
          return (
            <BirthdateInput
              initialValue='08/19/2019'
              {...props}
            />
          )
        },
        labelCopy: ' ',
        tid: 'birthdate-prefilled-tid',
        validators: [validateExists],
      },
    },
    onSubmit: async (formData) => {},
  }}
>
  {(api) => {
    const {
      field,
      getFormErrorMessage,
      getFormIsValid,
      getFormInteractedWith,
    } = api
    return (
      <div>
        <TitleLarge.Serif.Book500>
          Example Birthdate Prefilled
        </TitleLarge.Serif.Book500>

        <Spacer.H16 />

        {field('birthdate')}

        <Spacer.H16 />

        <Button.Medium.Black disabled={!getFormIsValid()} type="submit">
          Continue
        </Button.Medium.Black>
      </div>
    )
  }}
</Form>
```
