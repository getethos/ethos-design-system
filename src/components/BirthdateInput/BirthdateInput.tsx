import * as React from 'react'
import { DateInput } from '../DateInput'
import { PublicFormProps } from '../Form/Form'
import * as Validators from '../../validators/BirthdateInputValidator'
const { DATE_FORMATS } = Validators
type BirthdateInputProps = PublicFormProps & {
  dateFormat?: any
  disabled?: boolean
  allCaps?: boolean
  capitalize?: boolean
  labelCopy?: string
  initialValue?: string
  currentValue?: string
  currentError?: string
}
export const BirthdateInput: React.SFC<BirthdateInputProps> = (props) => {
  const {
    optional,
    dateFormat,
    allCaps,
    capitalize,
    labelCopy,
    validator,
    formChangeHandler,
    initialValue,
    currentValue,
    currentError,
    formTouched,
    setFieldTouched,
    ...restProps
  } = props
  return (
    <DateInput
      optional={optional}
      dateFormat={dateFormat}
      allCaps={allCaps}
      capitalize={capitalize}
      labelCopy={labelCopy}
      validator={validator}
      formChangeHandler={formChangeHandler}
      initialValue={initialValue}
      currentValue={currentValue}
      currentError={currentError}
      formTouched={formTouched}
      setFieldTouched={setFieldTouched}
      {...restProps}
    />
  )
}
export const BirthdateInputValidators = Validators
