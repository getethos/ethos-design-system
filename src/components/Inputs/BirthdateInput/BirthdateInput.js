import React, { useState } from 'react'
import PropTypes from 'prop-types'
import MaskedInput from 'react-text-mask'
import createAutoCorrectedDatePipe from 'text-mask-addons/dist/createAutoCorrectedDatePipe'

import { INPUT_MODES } from '../../../constants'
import dayjs from '../../../helpers/getDayjs.js'
import useErrorMessage from '../../../hooks/useErrorMessage.js'
import { InfoMessage } from '../../index'
import { InputLabel } from '../InputLabel'
import { Spacer } from '../../Spacer'
import * as Validators from './BirthdateInputValidator'

const {
  cleanse,
  DATE_FORMATS,
  dateMaskByFormat,
  dateRegexByFormat,
  dateStringMatchesFormat,
} = Validators

const PrivateBirthdateInput = (props) => {
  const {
    minAge,
    maxAge,
    name,
    dateFormat,
    allCaps,
    labelCopy,
    validator,
    onChange,
    forcedErrorMessage,
    value,
    ...restProps
  } = props

  const autoCorrectedDatePipe = createAutoCorrectedDatePipe('mm/dd/yyyy')
  const [touched, setTouched] = useState(false)

  const onBlur = (syntheticReactEvent) => {
    //   const cleansed = cleanse(syntheticReactEvent.target.value)
    setTouched(true)
  }

  //   // First check in valid format as that error takes priority
  //   let errMsg = dateStringMatchesFormat(cleansed, dateFormat)
  //   if (forcedErrorMessage) {
  //     setError(forcedErrorMessage)
  //   } else if (errMsg.length) {
  //     setError(errMsg)
  //   } else {
  //     // Now we let the validator validate the date range
  //     const conformedDate = dayjs(cleansed, dateFormat.toUpperCase())
  //     errMsg = validate(conformedDate)
  //     if (errMsg.length) {
  //       setError(errMsg)
  //     } else {
  //       // Passed all checks, reset error empty
  //       setError('')
  //     }
  //   }

  //   privateOnChange(syntheticReactEvent)
  // }

  // const privateOnChange = (syntheticReactEvent) => {
  //   if (!touched) return
  //   const cleansed = cleanse(syntheticReactEvent.target.value)
  //   const errMsg = dateStringMatchesFormat(cleansed, dateFormat)
  //   if (errMsg.length) {
  //     setError(errMsg)
  //   } else {
  //     setError('')
  //   }

  //   // TODO stop copying this code in all text inputs
  //   if (!!onChange) {
  //     onChange(syntheticReactEvent)
  //   }
  // }

  const err =
    touched && forcedErrorMessage ? (
      <>
        <Spacer.H8 />
        <InfoMessage.Text.Error>{forcedErrorMessage} </InfoMessage.Text.Error>
      </>
    ) : null

  return (
    <>
      <InputLabel
        name={'birthdate-auto-corrected'}
        labelCopy={labelCopy}
        allCaps={allCaps}
      />
      <MaskedInput
        mask={dateMaskByFormat[dateFormat]}
        pipe={autoCorrectedDatePipe}
        className={
          !!err ? 'BirthdateInput TextInput Error' : 'BirthdateInput TextInput'
        }
        type="text"
        inputMode={INPUT_MODES.NUMERIC}
        data-tid={restProps['data-tid']}
        guide={true}
        onBlur={onBlur}
        onChange={onChange}
        name="birthdate-auto-corrected"
        placeholder={dateFormat}
        keepCharPositions={true}
      />
      {err}
    </>
  )
}

PrivateBirthdateInput.PUBLIC_PROPS = {
  dateFormat: PropTypes.oneOf(DATE_FORMATS),
  'data-tid': PropTypes.string.isRequired,
  disabled: PropTypes.bool,
  allCaps: PropTypes.bool,
  minAge: PropTypes.number,
  maxAge: PropTypes.number,
  name: PropTypes.string.isRequired,
  labelCopy: PropTypes.string.isRequired,
  validator: PropTypes.func,
  onChange: PropTypes.func,
}

PrivateBirthdateInput.propTypes = {
  ...PrivateBirthdateInput.PUBLIC_PROPS,
}

PrivateBirthdateInput.defaultProps = {
  dateFormat: 'mm/dd/yyyy',
  labelCopy: 'Birthdate',
}

const BirthdateInputFactory = (privateProps) => {
  const PublicBirthdateInputComponent = (downstreamProps) => {
    return <PrivateBirthdateInput {...downstreamProps} {...privateProps} />
  }

  return PublicBirthdateInputComponent
}

export const BirthdateInput = BirthdateInputFactory()

export const BirthdateInputValidators = Validators
