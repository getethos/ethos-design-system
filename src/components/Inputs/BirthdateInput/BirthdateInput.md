The BirthdateInput takes a validator, but will also internally validate that the
date string corresponds with `dateFormat` (once the field has been `touched` similar
to [redux-form](https://redux-form.com/8.2.2/docs/api/field.md/#2-a-stateless-function))

```jsx
import { getMinMaxDateValidator } from './BirthdateInputValidator.js'
;<BirthdateInput
  name="birthDate"
  allCaps={true}
  labelCopy="Birthdate"
  data-tid="the-birthdate-input"
  minAge={20}
  maxAge={65}
  validator={(value) => {
    return getMinMaxDateValidator({
      minAge: 20,
      maxAge: 65,
      dateFormat: 'mm/dd/yyyy',
    })(value)
  }}
/>
```

```jsx
<BirthdateInput
  name="birthDate"
  allCaps={true}
  labelCopy="Birthdate"
  data-tid="the-birthdate-input"
  minAge={20}
  maxAge={65}
  forcedErrorMessage="Forced"
  validator={(value) => {}}
/>
```
