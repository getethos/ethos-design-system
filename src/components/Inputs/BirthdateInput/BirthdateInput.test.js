import React from 'react';
import { BirthdateInput } from './BirthdateInput';
import renderer from 'react-test-renderer';

// TODO: make a helper
// https://github.com/text-mask/text-mask/issues/427
jest.mock('react-text-mask', () => (props) => <input type="text" {...props} />)

describe('BirthdateInput', () => {
  it('default rendering', () => {
    const tree = renderer
      .create(
        <BirthdateInput
          dateFormat='mm/dd/yyyy'
          name="le-birthdate"
          allCaps={true}
          labelCopy="What is your birthdate?"
          data-tid='le-birthdate'
          validator={(thing) => {}}
          name="birthDate"
          minAge={20}
          maxAge={65}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
