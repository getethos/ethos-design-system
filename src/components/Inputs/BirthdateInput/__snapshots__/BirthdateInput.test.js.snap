// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BirthdateInput default rendering 1`] = `
Array [
  <label
    className="Caption Theinhardt Medium500 GrayPrimary AllCaps"
    htmlFor="birthdate-auto-corrected"
  >
    What is your birthdate?
  </label>,
  <div
    className="Spacer"
    style={
      Object {
        "height": 8,
      }
    }
  />,
  <input
    className="TextMaskedInput TextInput"
    data-tid="le-birthdate"
    disabled={false}
    guide={true}
    keepCharPositions={true}
    mask={
      Array [
        /\\\\d/,
        /\\\\d/,
        "/",
        /\\\\d/,
        /\\\\d/,
        "/",
        /\\\\d/,
        /\\\\d/,
        /\\\\d/,
        /\\\\d/,
      ]
    }
    name="birthdate-auto-corrected"
    onBlur={[Function]}
    onChange={[Function]}
    onPaste={[Function]}
    pipe={[Function]}
    placeholder="mm/dd/yyyy"
    type="tel"
    value=""
  />,
  "",
]
`;
